ray-operator/controllers/ray/rayjob_controller.go

```diff
@@ -888,6 +899,19 @@ func (r *RayJobReconciler) constructRayClusterForRayJob(rayJobInstance *rayv1.Ra
 		return nil, err
 	}
 
+	// Inject a submitter container into the head Pod in SidecarMode.
+	if rayJobInstance.Spec.SubmissionMode == rayv1.SidecarMode {
+		sidecar, err := getSubmitterContainer(ctx, rayJobInstance, rayCluster)
+		if err != nil {
+			return nil, err
+		}
+		rayCluster.Spec.HeadGroupSpec.Template.Spec.Containers = append(
+			rayCluster.Spec.HeadGroupSpec.Template.Spec.Containers, sidecar)
+		// Ensure the head Pod doesn't restart after sidecar termination
+		// For more details, see https://github.com/ray-project/kuberay/issues/3928#issuecomment-3187164736
```

Comment:
"For K8sJobMode, the submitter K8s Job relies on the K8s Job backoffLimit API to restart the submitter if it fails. This is mainly used to handle WebSocket connection failures caused by transient network issues. However, in SidecarMode, the submitter container shares the same network namespace as the Ray dashboard, so restarts are no longer needed."